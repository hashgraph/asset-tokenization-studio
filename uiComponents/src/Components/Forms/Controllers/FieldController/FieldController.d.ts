import type { FormErrorMessageProps as ChakraFormErrorMessageProps } from "@chakra-ui/react";
import type { ControllerFieldState } from "react-hook-form";
import React from "react";
export interface FieldControllerProps {
    children: React.ReactNode;
    fieldState: ControllerFieldState;
    showErrors?: boolean;
    errorMessageVariant: ChakraFormErrorMessageProps["variant"];
}
export declare const FieldController: ({ children, fieldState, errorMessageVariant, showErrors, }: FieldControllerProps) => JSX.Element;
