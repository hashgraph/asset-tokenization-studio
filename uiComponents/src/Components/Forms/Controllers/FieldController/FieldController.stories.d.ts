declare const meta: import("@storybook/types").ComponentAnnotations<import("@storybook/react/dist/types-0a347bb9").R, import("./FieldController").FieldControllerProps>;
export default meta;
export declare const Variant: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("./FieldController").FieldControllerProps>;
export declare const WithErrors: import("@storybook/types").AnnotatedStoryFn<import("@storybook/react/dist/types-0a347bb9").R, import("./FieldController").FieldControllerProps>;
