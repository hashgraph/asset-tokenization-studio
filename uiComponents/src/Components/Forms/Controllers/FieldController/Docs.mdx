import {
  Story,
  Meta,
  Title,
  ArgTypes,
  Primary,
  Stories,
  Source,
} from "@storybook/addon-docs";
import { Grid, GridItem, Text, Link, Center, Flex } from "@chakra-ui/react";
import * as FieldControllerStories from "./FieldController.stories";

<Meta of={FieldControllerStories} />

<Title />

<Primary />

## How to set up

FieldController is built with FormControl

## Props

<ArgTypes of={FieldControllerStories} />

## Stories

<Stories title="" />
