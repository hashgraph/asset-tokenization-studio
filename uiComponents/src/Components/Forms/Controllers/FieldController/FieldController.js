import { FormControl as ChakraFormControl, FormErrorMessage as ChakraFormErrorMessage, } from "@chakra-ui/react";
import React from "react";
export var FieldController = function (_a) {
    var _b;
    var children = _a.children, fieldState = _a.fieldState, errorMessageVariant = _a.errorMessageVariant, _c = _a.showErrors, showErrors = _c === void 0 ? true : _c;
    var isInvalid = !!(fieldState === null || fieldState === void 0 ? void 0 : fieldState.error);
    var message = (_b = fieldState === null || fieldState === void 0 ? void 0 : fieldState.error) === null || _b === void 0 ? void 0 : _b.message;
    return (React.createElement(ChakraFormControl, { isInvalid: isInvalid },
        children,
        showErrors && !!message && (React.createElement(ChakraFormErrorMessage, { "data-testid": "form-error-message", variant: errorMessageVariant }, message))));
};
