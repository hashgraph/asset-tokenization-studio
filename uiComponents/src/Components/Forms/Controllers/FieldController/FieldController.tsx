import type { FormErrorMessageProps as ChakraFormErrorMessageProps } from "@chakra-ui/react";
import {
  FormControl as ChakraFormControl,
  FormErrorMessage as ChakraFormErrorMessage,
} from "@chakra-ui/react";
import type { ControllerFieldState } from "react-hook-form";
import React from "react";

export interface FieldControllerProps {
  children: React.ReactNode;
  fieldState: ControllerFieldState;
  showErrors?: boolean;
  errorMessageVariant: ChakraFormErrorMessageProps["variant"];
}

export const FieldController = ({
  children,
  fieldState,
  errorMessageVariant,
  showErrors = true,
}: FieldControllerProps) => {
  const isInvalid = !!fieldState?.error;
  const message = fieldState?.error?.message;

  return (
    <ChakraFormControl isInvalid={isInvalid}>
      {children}
      {showErrors && !!message && (
        <ChakraFormErrorMessage
          data-testid="form-error-message"
          variant={errorMessageVariant}
        >
          {message}
        </ChakraFormErrorMessage>
      )}
    </ChakraFormControl>
  );
};
