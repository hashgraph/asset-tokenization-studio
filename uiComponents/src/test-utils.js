import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
