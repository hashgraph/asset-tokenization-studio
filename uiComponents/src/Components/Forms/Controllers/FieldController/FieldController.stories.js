var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React from "react";
import { FieldController } from "./FieldController";
import { Box } from "@chakra-ui/react";
var meta = {
    title: "Design System/Forms/Controllers/FieldController",
    component: FieldController,
    args: {},
    argTypes: {
        fieldState: {
            table: {
                disable: true
            },
            control: false
        }
    },
    parameters: {
        design: {
            type: "figma",
            url: "https://www.figma.com/file/YF8AvIq0MclfhGDvKgzypp/NEW---ioBricks-Library?node-id=1503%3A22829&t=4DeUdPmWLzP51oPO-4"
        },
        docs: {}
    }
};
export default meta;
var Template = function (args) {
    return (React.createElement(Box, { bg: "neutral.100", p: 4 },
        React.createElement(FieldController, __assign({}, args),
            React.createElement("input", { type: "text" }))));
};
export var Variant = Template.bind({});
Variant.args = {
    errorMessageVariant: "flushed",
    fieldState: {
        error: { type: "example", message: "This is an error" },
        invalid: true,
        isTouched: true,
        isDirty: true
    }
};
export var WithErrors = Template.bind({});
WithErrors.args = {
    fieldState: {
        error: { type: "example", message: "This is an error" },
        invalid: true,
        isTouched: true,
        isDirty: true
    }
};
