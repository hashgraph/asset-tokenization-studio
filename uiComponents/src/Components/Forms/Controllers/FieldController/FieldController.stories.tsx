import type { Meta, StoryFn } from "@storybook/react";
import React from "react";

import { FieldController } from "./FieldController";
import { Box } from "@chakra-ui/react";

const meta = {
  title: "Design System/Forms/Controllers/FieldController",
  component: FieldController,
  args: {},
  argTypes: {
    fieldState: {
      table: {
        disable: true,
      },
      control: false,
    },
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/YF8AvIq0MclfhGDvKgzypp/NEW---ioBricks-Library?node-id=1503%3A22829&t=4DeUdPmWLzP51oPO-4",
    },
    docs: {},
  },
} as Meta<typeof FieldController>;
export default meta;

const Template: StoryFn<typeof FieldController> = (args) => {
  return (
    <Box bg="neutral.100" p={4}>
      <FieldController {...args}>
        <input type="text" />
      </FieldController>
    </Box>
  );
};
export const Variant = Template.bind({});
Variant.args = {
  errorMessageVariant: "flushed",
  fieldState: {
    error: { type: "example", message: "This is an error" },
    invalid: true,
    isTouched: true,
    isDirty: true,
  },
};

export const WithErrors = Template.bind({});
WithErrors.args = {
  fieldState: {
    error: { type: "example", message: "This is an error" },
    invalid: true,
    isTouched: true,
    isDirty: true,
  },
};
