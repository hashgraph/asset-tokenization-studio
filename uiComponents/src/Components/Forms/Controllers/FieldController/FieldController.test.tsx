import { waitFor } from "@testing-library/react";
import type { UseFormReturn } from "react-hook-form";
import { useController, useForm } from "react-hook-form";
import type { FieldControllerProps } from "./FieldController";
import { FieldController } from "./FieldController";
import React from "react";
import { render } from "@/test-utils";
import { act } from "react-dom/test-utils";

type DefaultProps = Omit<FieldControllerProps, "fieldState">;
const fieldRequired = "This field is required";

describe(`<FieldController />`, () => {
  const defaultProps: DefaultProps = {
    showErrors: true,
    errorMessageVariant: "solid",
    children: <input type="text" data-testid="input" />,
  };

  let form: UseFormReturn;
  const RenderWithForm = (props: DefaultProps) => {
    const localForm = useForm({
      mode: "onChange",
      criteriaMode: "all",
    });
    form = localForm;
    const { control } = localForm;
    const { fieldState } = useController({
      name: "field",
      control,
      rules: { required: fieldRequired },
    });

    return <FieldController fieldState={fieldState} {...props} />;
  };

  const factoryComponent = (props: DefaultProps = defaultProps) => {
    return render(<RenderWithForm {...props} />);
  };

  test("Component renders correctly", () => {
    const component = factoryComponent();

    expect(component.asFragment()).toMatchSnapshot();
    expect(component.getByTestId("input")).toBeInTheDocument();
  });

  test("Should show errors when showErrors=true", async () => {
    const component = factoryComponent();
    act(() => {
      form.setError("field", { type: "invalid", message: fieldRequired });
    });

    await waitFor(() => {
      const FormErrorMessage = component.getByTestId("form-error-message");
      expect(FormErrorMessage).toHaveTextContent(fieldRequired);
    });

    expect(component.asFragment()).toMatchSnapshot("WithInvalidMessage");
  });

  test("Should hide errors when showErrors=false", async () => {
    const component = factoryComponent({ ...defaultProps, showErrors: false });
    act(() => {
      form.setError("field", { type: "invalid", message: fieldRequired });
    });

    await waitFor(() => {
      expect(component.queryByTestId("form-error-message")).toBeNull();
    });

    expect(component.asFragment()).toMatchSnapshot(
      "WithInvalidNoFormErrorMessage"
    );
  });
});
