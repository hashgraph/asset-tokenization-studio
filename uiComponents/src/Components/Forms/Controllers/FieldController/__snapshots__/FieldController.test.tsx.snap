// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldController /> Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-form-control css-1kxonj9"
    role="group"
  >
    <input
      data-testid="input"
      type="text"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<FieldController /> Should hide errors when showErrors=false: WithInvalidNoFormErrorMessage 1`] = `
<DocumentFragment>
  <div
    class="chakra-form-control css-1kxonj9"
    role="group"
  >
    <input
      data-testid="input"
      type="text"
    />
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<FieldController /> Should show errors when showErrors=true: WithInvalidMessage 1`] = `
<DocumentFragment>
  <div
    class="chakra-form-control css-1kxonj9"
    role="group"
  >
    <input
      data-testid="input"
      type="text"
    />
    <div
      aria-live="polite"
      class="chakra-form__error-message css-jm4k09"
      data-testid="form-error-message"
      id="field-:r1:-feedback"
    >
      This field is required
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
